Doc xrt exe

uday610 opened the pull request:
@sonals , @stsoe , please review doc for kernel execution models

gbuildx commented:
Can one of the admins verify this patch?

sonals commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,105 @@
+2. There is no need to start the kernel by *clEnqueueTask* or *clEnqueueNDRangeKernel* from the host. 
+3. Host comminucates with a continuously running kernel by the stream read and write requests, if necessary. 
+4. Dont use clSetKernelArg to pass scalar argument to ap_ctrl_none kernel, only use *xclRegWrite* (API to be implemented in 19.2) API. 
+
```
On item 4 above for AP_CTRL_NONE, we should suggest new API xclRegWrite/xclRegRead and also clarify that this requires exclusive context via xclOpenContext() on the CU.

uday610 replied:
Thanks, Added the following:

In general, for the purpose of read and write register values from AXI4-Lite Slave interface, the new APIs *xclRegRead/xclRegWrite* are recommended (replacing obsolated APIs *xclRead/xclWrite*). As shown in the code sample below, these APIs requires exclusive CU context reservation via API *xclOpenContext*. 

.. code-block:: c

   xclOpenContext(device_handle, xclbin_id, cu_index, false);
   xclRegRead(device_handle, cu_index, offset, &data);
   xclRegWrite(device_handle, cu_index, offset, data);
   xclCloseContext(device_handle, xclbin_id, cu_index);

sonals commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,105 @@
+
+|START1|=>|START2|=>|START3|=>|DONE1|=>|START4|=>|DONE2|=>|START5|=>|DONE3|=>|DONE4|=>|DONE5|
+
+
```
We should call emphasize that AP_CTRL_CHAIN implies all requests are serviced _in order_.

uday610 replied:
Good point, added the following:
|START1|=>|START2|=>|START3|=>|DONE1|=>|START4|=>|DONE2|=>|START5|=>|DONE3|=>|DONE4|=>|DONE5|

**Note:** As noted in the above sequence, the AP_CTRL_CHAIN only applicable when the kernel produces the outputs for the pending requests in-order. Kernel servicing the requests out-of-order cannot be supported by AP_CTRL_CHAIN execution model.

uday610 commented:
> This is very good!
> We should also describe how are these kernels types are identified by XRT -- xclbin IP LAYOUT specifics on how is this information captured.

Ok, I added the following: 

The IP_LAYOUT section of the kernel xclbin metadata contains the kernel execution model information. The xclbinutil utility command can be used to retrieve the information from a xclbin file. 

.. code-block::

   xclbinutil --info --input binary_container_1.xclbin --dump-section IP_LAYOUT:json:ip_layout_info.json

.. code-block:: json
   {
    "ip_layout": {
        "m_count": "5",
        "m_ip_data": {
            "ip_data": {
                "m_type": "IP_KERNEL",
                "m_int_enable": "0",
                "m_interrupt_id": "0",
                "m_ip_control": "AP_CTRL_HS",

keryell commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,136 @@
+
+The list of primary supported excution models are:
+
+1. AP_CTRL_HS
```
A lot of missing backquote around `AP_CTRL_HS` and everywhere.

keryell commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,136 @@
+2. AP_CTRL_CHAIN
+3. AP_CTRL_NONE
+
+The IP_LAYOUT section of the kernel xclbin metadata contains the kernel execution model information. The xclbinutil utility command can be used to retrieve the execution model information from a xclbin file. 
```
`IP_LAYOUT` `xclbin` `xclbinutil` `.xclbin`

keryell commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,136 @@
+
+**Important points to remember**
+
+1. Consider a kernel with AP_CTRL_NONE only when it has no Memory input and output.
```
memory

keryell commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,136 @@
+**Important points to remember**
+
+1. Consider a kernel with AP_CTRL_NONE only when it has no Memory input and output.
+2. There is no need to start the kernel by *clEnqueueTask* or *clEnqueueNDRangeKernel* from the host. 
```
*clEnqueueTask* → `clEnqueueTask` and so on, since it is an API and not a concept

keryell commented on src/runtime_src/doc/toc/xrt_kernel_executions.rst:
```diff
@@ -0,0 +1,136 @@
+
+1. Consider a kernel with AP_CTRL_NONE only when it has no Memory input and output.
+2. There is no need to start the kernel by *clEnqueueTask* or *clEnqueueNDRangeKernel* from the host. 
+3. Host comminucates with a continuously running kernel by the stream read and write requests, if necessary. 
```
comminucates → communicates

You should configure your environment to use a spell-checker...

uday610 commented:
Thanks @keryell , 

1) I changed those images to remove shadows and dots. 
2) Also corrected few typos and used backquotes for OpenCL commands instead of italics. 

I am not big fan of adding backquotes for every FPGA related words such as xclbin, IP_LAYOUT etc, bI feel highlighting too many words ruins the ease of reading the articles. Generally I prefer to highlight (or make it bold) for a newly introduced word, so I am keeping highlights for ap_ctrl_hs, ap_ctrl_chain etc and also for commands like clEnqueueTask etc.
